src/Compilers/Core/Portable/InternalUtilities/EnumerableExtensions.cs

```diff
@@ -382,18 +382,56 @@ public static T[] AsArray<T>(this IEnumerable<T> source)
             return builder.ToImmutableAndFree();
         }
 
-        public static ImmutableArray<TResult> SelectManyAsArray<TSource, TResult>(this IReadOnlyCollection<TSource>? source, Func<TSource, IEnumerable<TResult>> selector)
+        public static ImmutableArray<TResult> SelectManyAsArray<TSource, TResult>(this IEnumerable<TSource>? source, Func<TSource, IEnumerable<TResult>> selector)
         {
             if (source == null)
                 return ImmutableArray<TResult>.Empty;
 
+            var builder = ArrayBuilder<TResult>.GetInstance();
+            foreach (var item in source)
+                builder.AddRange(selector(item));
+
+            return builder.ToImmutableAndFree();
+        }
+
+        public static ImmutableArray<TResult> SelectManyAsArray<TItem, TArg, TResult>(this IEnumerable<TItem>? source, Func<TItem, TArg, IEnumerable<TResult>> selector, TArg arg)
```

Comment:
I really dislike our "put the `arg` parameter last" convention. It is actively bad for using these methods, as it means the compiler can't infer the type of the lambda parameter until after you've already typed it out, or until you break the typing flow to add the trailing parameter and then move back to the lambda body. Can we move `arg` before `selector`?

Reply:
@333fred i did that in the past and got asked to keep things consistent :)

My preference would be to have them all behave the same way, and then do a full switch on everything if we get buyoff.  Thanks!

Reply:
Unfortunately, I think there are public APIs using this pattern.

Reply:
What would be really cool is compiler support for the pattern. If I use the overload that doesn't take TArg but the lambda captures any variables it would automatically emit the other overload, passing in the captured state.

Reply:
> Unfortunately, I think there are public APIs using this pattern.

Frankly, I don't think public apis should prevent us from making internal APIs usable. This pattern is extremely ergonomic, and I do not want it to continue to spread.

Reply:
@333fred could i make a followup pr that then consistently moves these all at once?

Reply:
If you must, but I will continue to register my complaint against this pattern, even at the cost of inconsistency.

Reply:
I'd actually prefer to keep the patter as is and implement compiler feature as suggested above that makes this discussion moot.

Reply:
That's not a compiler feature, that's a language feature, one otherwise known as `original`/`replace` 🙂. Fixing up our overloads is significantly simpler.

Reply:
original/replace is very generic. This can be a compiler feature that only handles closures for methods annotated by some attribute. The generalization can be added later on.

Reply:
let's take the dsicussion elsewhere :D

Reply:
> This can be a compiler feature that only handles closures for methods annotated by some attribute.

That is still a massive language feature.

Reply:
It's requested yet another time at https://github.com/dotnet/runtime/issues/94654. I'd like to see effort allocated for this.